pkg/controller/core.oam.dev/v1alpha2/application/application_controller.go

```diff
@@ -241,13 +224,66 @@ func (r *Reconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 	return ctrl.Result{}, r.UpdateStatus(ctx, app)
 }
 
-// if any finalizers newly registered, return true
-func registerFinalizers(app *v1beta1.Application) bool {
-	if !meta.FinalizerExists(&app.ObjectMeta, resourceTrackerFinalizer) && app.Status.ResourceTracker != nil {
-		meta.AddFinalizer(&app.ObjectMeta, resourceTrackerFinalizer)
-		return true
+func (r *Reconciler) handleFinalizers(ctx context.Context, app *v1beta1.Application) (bool, error) {
+	if app.ObjectMeta.DeletionTimestamp.IsZero() {
+		// NOTE Because resource tracker is cluster-scoped resources, we cannot garbage collect them
+		// by setting application(namespace-scoped) as their owner. So we must delete all
+		// resource trackers through app controller's finalizer logic.
```

Comment:
This explanation should be moved to the function signature

Reply:
Got it.